062-quotas-management.md

```diff
@@ -0,0 +1,163 @@
+   Having a dedicated API for quota configuration would make it much easier to have the operator configure the client automatically.
+
+## Proposal
+
+This proposal suggests to add a new `quotas` API to the `Kafka` custom resource.
+It will be inside the `.spec.kafka` section.
+And it will allow configuring different types of quotas.
+As many existing Strimzi APIs, it will use the `type` field to differentiate the quota plugin that will be used.
+
+This proposal suggests to initially add support for two Quota plugins.
+1. The build-in Kafka plugin
+2. The Strimzi Quotas plugin
+
+The first supported plugin - the Kafka's built-in plugin - will be configured as `type: kafka`.
+The API will allow to configure the default user quotas for this plugin in the same format as the `KafkaUser` API allows it today:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: kafka
+      producerByteRate: 1048576
+      consumerByteRate: 2097152
+      requestPercentage: 55
+      controllerMutationRate: 50
+    # ... 
+  # ...
+```
+
+Based on this configuration, the operator will:
+* Configure the default quota plugin (make sure the `client.quota.callback.class` option is not set)
+* Set the default user quotas
+* When the `type: kafka` quotas APi will be set, but one or more of fields with the individual quotas will be missing, the operator will make sure that given default quota is not set.
+* The other configuration options for the built-in Kafka plugin can be configured by the user in the `.spec.kafka.config` section in the `Kafka` CR.
+
+The second supported plugin - the Strimzi Quotas plugin - will be configured as `type: strimzi`.
+The API will allow to configure the storage limits at which the plugin will start limiting the producers in order to prevent the disk from getting full.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: strimzi
+      produceQuota: 1000000
+      fetchQuota: 1000000
+      minAvailableBytesPerVolume: 5368709120
+      excludedPrincipals:
+        - my-special-user
+    # ... 
+  # ...
+```
+
+The `type: strimzi` quotas API will support the following fields:
+* `produceQuota` for setting the default produce quota
+* `fetchQuota` for setting the default fetch quota
+* `minAvailableBytesPerVolume` for setting the minimal required free space per volume as number of bytes (this field can be configured only when `minAvailableRatioPerVolume`is not set)
+* `minAvailableRatioPerVolume` for setting the minimal required free space per volume as percentage of the overall volume size (this field can be configured only when `minAvailableBytesPerVolume`is not set)
+* `excludedPrincipals` for setting s list of users to be excluded from the applied quotas
```

Comment:
Is this an exclusive list or does Strimzi implicitly know of other super user type user which should not be blocked and can thus add them automagically?

nit typo: 
```suggestion
* `excludedPrincipals` for setting a list of users to be excluded from the applied quotas
```

Reply:
The configuration mostly copies the plugin configuration as described in the README. So this expects the user to specify any users for that it would not apply (not entirely sure what was the motivation was for this).

If some internal users need to be excluded as well, please let me know which and why and I can add it to the proposal. We should probably also add it to the plugin README?

Reply:
No, I'm not aware of any specific user which need to be excluded. I was just wanting to clarify if the values in here were expected to be an exclusive list or merged with a Strimzi internal one. From your answer its clear it's exclusive. 

The motivation is to ensure that configured admin clients could still access the cluster e.g. to persist more up to date readings for disk usage in the cluster (from the earlier iteration of the plug-in).

Reply:
> The motivation is to ensure that configured admin clients could still access the cluster e.g. to persist more up to date readings for disk usage in the cluster (from the earlier iteration of the plug-in).

Well, Strimzi has _admin clients_. But they usually do not produce or fetch messages to/from topics. So they should be mostly unaffected?

That said, there are some exceptions ...
* Replication => I assume that does not have to be excluded, as the plugin should not affect it?
* Cruise Control => Do we have any reasons to have it excluded? For doing a rebalance with Cruise Control if the disk space issue caused by an imbalance in disk usage?
* Bidirectional Topic Operator => I would not expect it to produce enough data to fill in the disk. But not sure we want it excluded. Maybe to allow users to reconfigure retention using it? This problem will go away once BTO is removed since UTO does not have any topic.

TBH, I do not know the plugin details, I'm just trying to bring it to some end, but really rely on your input here 😉

Reply:
The Strimzi plug-in only applies limits to produce requests to topics so all other operations are un-affected.

So:

- `Replication` is un-affected. 
- `Cruise Control` might be worth so that it can refresh its metrics even when external users are blocked. (If I am right in saying it has a topic to store its metrics)
- `BTO` Would need to be included in the allow list so that it could operate while the cluster is blocked.
- `Canary` would be an other example of something to potentially exclude by default (though that might be better handled in the canary documentation).

Reply:
Canary is not deployed through the Kafka CR. So that is something users will need to exclude them self using the option.

Reply:
I updated the proposal accordingly.